### ISSUE TITLE
How to find lidar_segmentation label for a particular 3d point in nuScenes?

### ISSUE THREAD

**Issue by daeunni** [nutonomy/nuscenes-devkit#786]
Hi, thx for your nice works! :) 
i have questions about **merging** nuScenes + nuScenes_lidaraseg. 

First of all, I'm using 9 b-box parameters made by below. 
```
            nusc = NuScenes(version=self.dataset_cfg.VERSION, dataroot=str(self.root_path), verbose=True)
            annotations = [nusc.get('sample_annotation', token) for token in sample['anns']]

            # the filtering gives 0.5~1 map improvement
            num_lidar_pts = np.array([anno['num_lidar_pts'] for anno in annotations])
            num_radar_pts = np.array([anno['num_radar_pts'] for anno in annotations])
            mask = (num_lidar_pts + num_radar_pts > 0)

            locs = np.array([b.center for b in ref_boxes]).reshape(-1, 3)     # box center! 
            dims = np.array([b.wlh for b in ref_boxes]).reshape(-1, 3)[:, [1, 0, 2]]  # wlh == > dxdydz (lwh)
            velocity = np.array([b.velocity for b in ref_boxes]).reshape(-1, 3)
            rots = np.array([quaternion_yaw(b.orientation) for b in ref_boxes]).reshape(-1, 1)

            names = np.array([b.name for b in ref_boxes])
            tokens = np.array([b.token for b in ref_boxes])

            gt_boxes = np.concatenate([locs, dims, rots, velocity[:, :2]], axis=1)
```

And i think that the meaning of these 9 box parameters are as follows.
`gt_boxes = center point(x, y, z) + length + width + height + rotations + velocity`

Then, I want to know **the lidar segmentation value** corresponding to the (x, y, z) center point using nuScenes_lidarseg dataset in particular scene. 
In other words, I want to know **what segmentation label this 3D b-box's center point** corresponds to.

Can you help me to solve this work?
Best,

**Comment by whyekit-motional** [comment]
@daeunni the semantic label of the center of a given 3D box (`b.center` in your above code) would generally be the same as that of the class of the 3D box itself (`b.name` in your above code)

**Comment by daeunni** [comment]
Thanks for quick apply! :) 

Btw, i wonder what segmentation label a particular point has in a particular scene. 

For example, if there is a pointcloud segmentation(panoptic) of a scene with an "A" token, I want to know what the segmentation label of a particular (x,y,z) point is in "A" scene.

How can i get this?👍🏻

**Comment by whyekit-motional** [comment]
@daeunni I think you mean that you want to get the label of a particular (x, y, z) point in a sample (not scene, as points can shift across a scene)

To get the labels (both panoptic and lidarseg) for a particular point in a sample, here's how I would approach it:
```
import os.path as osp

import numpy as np

from nuscenes.nuscenes import NuScenes
from nuscenes.utils.data_classes import LidarPointCloud


# Load the nuScenes dataset (mini-split, in this case).
nusc = NuScenes(version='v1.0-mini', dataroot='/data/sets/nuscenes', verbose=False)


# Load some random sample token.
sd_record = nusc.get('sample', 'ca9a282c9e77460f8360f564131a8af5')
ref_sd_token = sd_record['data']['LIDAR_TOP']
ref_sd_record = nusc.get('sample_data', ref_sd_token)

# Load pointcloud.
pcl_path = osp.join(nusc.dataroot, ref_sd_record['filename'])
pc = LidarPointCloud.from_file(pcl_path)
points = pc.points.T

# Load panoptic labels.
panoptic_labels_filename = osp.join(nusc.dataroot, nusc.get('panoptic', ref_sd_token)['filename'])
panoptic_labels = np.load(panoptic_labels_filename)['data']

# Load panoptic labels.
lidarseg_labels_filename = osp.join(nusc.dataroot, nusc.get('lidarseg', ref_sd_token)['filename'])
lidarseg_labels = np.fromfile(lidarseg_labels_filename, dtype=np.uint8)

assert points.shape[0] == panoptic_labels.shape[0] == lidarseg_labels.shape[0]

# Get the labels of a desired point.
point_idx = 13579
print(f'Point at location {points[point_idx][:3]} has lidarseg label {lidarseg_labels[point_idx]} and panoptic label {panoptic_labels[point_idx]}')
```

**Comment by daeunni** [comment]
Thank you very much for your quick response. 😄
the above code seems to be able to find a segmentation label in the scene **only when a specific point_index exists.**

btw, how do i find a segmentation label that corresponds to a point when there is **no point_index** (simply assuming that i know just particular (x, y, z) coordinate)?

Thank you for being so helpful to me 👍🏻

**Comment by whyekit-motional** [comment]
@daeunni if there isn't a point, there wouldn't be an annotated label, haha

But if you really wanted to do such a thing, one naive way would be to: 

1. Check if a co-ordinate (x, y, z) lies within a bounding box
2. If the co-ordinate (x, y, z) does lies within a bounding box, then assign the class of the box to that co-ordinate

Note that there would be drawbacks to this method:

- As a bounding box is a coarse way of annotating an object, it might be possible that the co-ordinate (x, y, z) does not lie on the object itself, but is in the space surrounding the object (but still bounded by the box)
- You will not be able to get the class of a co-ordinate (x, y, z) which corresponds to a stuff class (e.g. vegetation, drivable surface, etc.) since these cannot be annotated with boxes

Feel free to share if you eventually develop something better!

**Comment by whyekit-motional** [comment]
Another possible method I can think of is to use a "nearest neighbors" approach, where you could find the N nearest annotated points to a given co-ordinate (x, y, z) and then assign the mode of the classes to the co-ordinate

One drawback of this approach would be, of course, the need to tune the value of N
